pkg/cvo/availableupdates_test.go

```diff
@@ -0,0 +1,306 @@
+	return nil, errors.NewNotFound(schema.GroupResource{Group: "", Resource: "configmap"}, name)
+}
+
+type fakeConditionRegistry struct{}
+
+func (f fakeConditionRegistry) Register(string, clusterconditions.Condition) {
+	panic("implement me")
+}
+
+func (f fakeConditionRegistry) PruneInvalid(ctx context.Context, matchingRules []configv1.ClusterCondition) ([]configv1.ClusterCondition, error) {
+	return matchingRules, nil
+}
+
+type clusterConditionRuleType string
+
+const (
+	ruleTypeAlways clusterConditionRuleType = "Always"
+	ruleTypePromQL clusterConditionRuleType = "PromQL"
+)
+
+type clusterConditionRuleFakePromql string
+
+const (
+	evalToYes   clusterConditionRuleFakePromql = "YES"
+	evalToNo    clusterConditionRuleFakePromql = "NO"
+	errorOnEval clusterConditionRuleFakePromql = "ERROR"
+	skip        clusterConditionRuleFakePromql = "SKIP"
+)
+
+func (f fakeConditionRegistry) Match(ctx context.Context, matchingRules []configv1.ClusterCondition) (bool, error) {
+	for _, rule := range matchingRules {
+		switch clusterConditionRuleType(rule.Type) {
+		case ruleTypeAlways:
+			return true, nil
+		case ruleTypePromQL:
+			switch clusterConditionRuleFakePromql(rule.PromQL.PromQL) {
+			case evalToYes:
+				return true, nil
+			case evalToNo:
+				return false, nil
+			case errorOnEval:
+				return false, fmt.Errorf("ERROR")
+			case skip:
+				continue
+			default:
+				panic("This fake works only with YES, NO, ERROR, and SKIP")
+			}
+		default:
+			panic("This fake works only with Always and PromQL risks")
+		}
+	}
+	return false, nil
+}
+
+func osusWithSingleConditionalEdge(from, to string, ruleType clusterConditionRuleType, promql clusterConditionRuleFakePromql) ([]configv1.ConditionalUpdate, *httptest.Server) {
+	osus := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Fprintf(w, `{
+  "nodes": [{"version": "%s", "payload": "payload/%s"}, {"version": "%s", "payload": "payload/%s"}],
+  "conditionalEdges": [
+    {
+      "edges": [{"from": "%s", "to": "%s"}],
+      "risks": [
+        {
+          "url": "https://example.com/%s",
+          "name": "FourFiveSix",
+          "message": "Four Five Five is just fine",
+          "matchingRules": [{"type": "%s", "promql": { "promql": "%s"}}]
+        }
+      ]
+    }
+  ]
+}
+`, from, from, to, to, from, to, to, ruleType, promql)
+	}))
+
+	updates := []configv1.ConditionalUpdate{
+		{
+			Release: configv1.Release{Version: to, Image: "payload/" + to},
+			Risks: []configv1.ConditionalUpdateRisk{
+				{
+					URL:     "https://example.com/" + to,
+					Name:    "FourFiveSix",
+					Message: "Four Five Five is just fine",
+					MatchingRules: []configv1.ClusterCondition{
+						{
+							Type:   "PromQL",
+							PromQL: &configv1.PromQLClusterCondition{PromQL: string(evalToYes)},
+						},
+					},
+				},
+			},
+			Conditions: []metav1.Condition{
+				{
+					Type:    "Recommended",
+					Status:  metav1.ConditionFalse,
```

Comment:
Here we hard-code the field `Type` to the value of `"PromQL"`, the field `PromQL` to the value of `string(evalToYes)`, and the field `Status` to the value of `metav1.ConditionFalse` even though they should be set by the function's parameters `ruleType`, `promql`, and by the expected value for the `Status` depending on the `Match` result.

Reply:
I think Petr has things the way he does so he can excercise "what we already know about a conditional update for a particular target, and the new update service response comes in with new information for that same image?".  So maybe we move from `PromQL` to `Always` like openshift/cincinnati-graph-data#3590.  The test suite should confirm that `lastTransitionTime` is being preserved if we transition from "outgoing PromQL matched and so does the incoming `Always`" and that the transition time resets if we transition from "outgoing PromQL did not match but the incoming `Always` does".

Reply:
> I think Petr has things the way he does so he can excercise "what we already know about a conditional update for a particular target, and the new update service response comes in with new information for that same image?". 

Oh, I didn't think of it like that, good point. Any of the options seems reasonable.

Reply:
Unfortunately I cannot claim the intent that Trevor assumes. The truth is that I finished the testing code knowing it is probably a bit rough but I called it day thinking it's not worth the effort to polish it entirely. David is right that this method pretends to be a reusable configurable helper that can prepare fixtures for different situations but in reality it does not. I'll spend some time looking at how I could clean it up even further, maybe taking other Trevor's review items into account as well.

Thanks for being thorough!